File: src/monitor.rs
use std::sync::atomic::AtomicU64;
use std::sync::atomic::Ordering::{Relaxed, SeqCst};
use std::sync::Arc;
use std::time::{Duration, Instant};

use crossbeam::channel::{unbounded, RecvTimeoutError, Sender};
use histogram::Histogram;

#[derive(Clone)]
pub struct Monitor {
    op_count: Arc<AtomicU64>,
    duration_tx: Sender<(Duration, u64)>,
}

impl Monitor {
    pub fn start_monitoring(report_interval_ms: u64) -> Self {
        let (tx, rx) = unbounded::<(Duration, u64)>();
        let count = Arc::new(AtomicU64::new(0));
        let count0 = count.clone();
        std::thread::spawn(move || {
            let mut now = Instant::now();
            let mut next_report_time = now + Duration::from_millis(report_interval_ms);
            let mut histogram = Histogram::new();
            let mut last_count = 0;

            loop {
                now = Instant::now();
                if now >= next_report_time {
                    let count = count0.load(SeqCst);
                    let qps = (count - last_count) as f64 / (report_interval_ms as f64 / 1_000.0);
                    let p50 =
                        histogram.percentile(50.0).unwrap_or(9999999999999) as f64 / 1_000_000.0;
                    let p90 =
                        histogram.percentile(90.0).unwrap_or(9999999999999) as f64 / 1_000_000.0;
                    let p99 =
                        histogram.percentile(99.0).unwrap_or(9999999999999) as f64 / 1_000_000.0;
                    let p999 =
                        histogram.percentile(99.9).unwrap_or(9999999999999) as f64 / 1_000_000.0;
                    let minimum = histogram.minimum().unwrap_or(9999999999999) as f64 / 1_000_000.0;
                    let mean = histogram.mean().unwrap_or(9999999999999) as f64 / 1_000_000.0;
                    let maximum = histogram.maximum().unwrap_or(9999999999999) as f64 / 1_000_000.0;
                    let stddev = histogram.stddev().unwrap_or(9999999999999) as f64 / 1_000_000.0;

                    println!("QPS: {:.2} | Percentiles (ms): p50: {:.2} p90: {:.2} p99: {:.2} p999: {:.2} | Latency (ms): Min: {:.2} Avg: {:.2} Max: {:.2} StdDev: {:.2}",
                             qps, p50, p90, p99, p999, minimum, mean, maximum, stddev,
                    );

                    last_count = count;
                    next_report_time = now + Duration::from_millis(report_interval_ms);
                    histogram = Histogram::new();
                }

                match rx.recv_timeout(next_report_time - now) {
                    Ok((value, count)) => histogram
                        .increment_by(value.as_nanos() as _, count)
                        .unwrap(),
                    Err(RecvTimeoutError::Disconnected) => break,
                    _ => {}
                }
            }
        });

        Self {
            op_count: count,
            duration_tx: tx,
        }
    }

    pub fn incr_count(&self, count: u64) {
        self.op_count.fetch_add(count, Relaxed);
    }

    pub fn add_response_time(&self, duration: Duration) {
        let _ = self.duration_tx.send((duration, 1));
    }
}

File: src/main/scala/Monitor.scala
import com.twitter.util.Future
import org.HdrHistogram.Histogram
import org.slf4j.LoggerFactory

import java.util.concurrent.atomic.{AtomicBoolean, AtomicLong}
import java.util.concurrent.{LinkedTransferQueue, TimeUnit}

class Monitor {
  val opCount = new AtomicLong(0)
  val channel = new LinkedTransferQueue[(Long, Long)]
  val stopNotifier = new AtomicBoolean(false)
  var thread: Thread = _

  def incrCount(count: Long): Unit = {
    opCount.addAndGet(count)
  }

  def addResponseTime(nanos: Long, count: Long): Unit = {
    channel.put((nanos, count))
  }

  def stop(): Unit = {
    stopNotifier.set(true)
    thread.join()
  }
}

object Monitor {
  def startMonitoring(reportIntervalMillis: Long): Monitor = {
    val monitor = new Monitor
    val thread = new Thread {
      override def run(): Unit = {
        val logger = LoggerFactory.getLogger(Monitor.getClass)
        val histogram = new Histogram(3600000000000L, 3)
        var now = System.currentTimeMillis()
        var nextReportTime = now + reportIntervalMillis
        var lastCount = 0L

        while (!monitor.stopNotifier.get()) {
          now = System.currentTimeMillis()
          if (now > nextReportTime) {
            val count = monitor.opCount.get()
            val qps =
              (count - lastCount).toDouble / (reportIntervalMillis.toDouble / 1000)
            val p50 = histogram.getValueAtPercentile(50.0).toDouble / 1000000
            val p90 = histogram.getValueAtPercentile(90.0).toDouble / 1000000
            val p99 = histogram.getValueAtPercentile(99.0).toDouble / 1000000
            val p999 = histogram.getValueAtPercentile(99.9).toDouble / 1000000
            val min = histogram.getMinValue.toDouble / 1000000
            val avg = histogram.getMean / 1000000
            val max = histogram.getMaxValue.toDouble / 1000000
            val stddev = histogram.getStdDeviation / 1000000

            logger.info(
              f"QPS: $qps%.2f | Latency (ms): p50 $p50%.2f, p90 $p90%.2f, p99 $p99%.2f, p999 $p999%.2f, Min $min%.2f, Avg $avg%.2f, Max $max%.2f, Stddev $stddev%.2f"
            )

            lastCount = count
            nextReportTime = now + reportIntervalMillis
            histogram.reset()
          }

          val pair =
            monitor.channel.poll(nextReportTime - now, TimeUnit.MILLISECONDS)
          if (pair != null) {
            histogram.recordValueWithCount(pair._1, pair._2)
          }
        }
      }
    }
    monitor.thread = thread
    thread.start()
    monitor
  }

  implicit class MonitoredFuture[A](future: Future[A]) {
    def monBy(monitor: Monitor): Future[A] = future.monBy(monitor, 1)

    def monBy(monitor: Monitor, count: Long): Future[A] = {
      for {
        startNanos <- {
          monitor.incrCount(count)
          Future.value(System.nanoTime())
        }
        v <- future
        _ <- {
          monitor.addResponseTime(System.nanoTime() - startNanos, count)
          Future.value(())
        }
      } yield v
    }
  }
}
